### ISSUE TITLE
Need option to manually reduce stock on order level

### ISSUE THREAD

**Issue by melindahelt** [woocommerce/woocommerce#21754]
zd-1530164

In WooCommerce 3.4.7, the ability to manually reduce stock within the order was available:

![](https://cld.wthms.co/5t4nKR+)Image Link: https://cld.wthms.co/5t4nKR

When following the above process, and saving the order (keeping order in Payment Pending Status), stock is reduced (per order notes, outside of the screenshot).

That feature was removed in WooCommerce 3.5 and order stock is automatically updated only after the order is set to processing (after payment received).

While auto-updating is good and saves a step, this takes the ability to change stock on the order level out of the hands of store owners.

In this case, a store owner needs to make sure products do not go out of stock while waiting for payment (which, happens after items are received in this case), but could happen often in Wholesale arrangements where more time is needed than is covered in hold stock options.

**Comment by Xenokorp** [comment]
Hi.
I'm the person that asked Melinda about bringing that useful feature back.
Let me know if I can help explaining further.

**Comment by besmarques** [comment]
Hi,
Still waiting for this feature to return

**Comment by Xenokorp** [comment]
Thanks for the +1, much appreciated you weight in.

**Comment by rwaynick** [comment]
Also requested in 5737446-hc.

**Comment by Xenokorp** [comment]
No idea what it means, but it seems positive so we get that back :)

**Comment by Littleiskra** [comment]
Hello all. Woo Commerce support direct me to this thread. I really wont this useful feature back, too. :)

**Comment by Xenokorp** [comment]
+1 : all orders I shipped yesterday got all the stock reduced twice (once by me when placed, once by the system upon completion). 300 products to adjust the stock for afterwards... A TOTAL PAIN!

**Comment by negapo** [comment]
I cannot understand why this feature was removed. WooCommerce doesn't properly manage stock increases/reductions on editing an order on the backend when we add/remove products (an old discussion between developers and shop owners). Without this feature this is still more absurd, there is no way to properly edit an order, we have to add/remove products and then go to to product page, find the variation if is a variable product and then adjust inventory. Absurd.

**Comment by zzbeck** [comment]
Also waiting for this feature to return.

**Comment by ianulpindo** [comment]
Please bring it back/fix it! I always need to edit an order due to customer requests. Thank you!

**Comment by kloon** [comment]
Related #21837

**Comment by ben72** [comment]
+1 to get this back! I use this for multiple things like:
1. When picking the order it's a great way to keep track of progress
2. Put a returned item back into stock when for instance changing to another size and get the action logged at the same time.
3. Remove an item from stock if replacing a faulty item and get the action logged at the same time.
4. Put items back in stock when pre-paid order is canceled.

**Comment by negapo** [comment]
Any rational why this was removed?
In a perfect world (very reachable) we wouldn't need this. Here is the expected behavior of every shop owner on stock changes: 

Principles:
First principle, stock changes should be the same if an order has changed status in a manual manner (thru the admin panel) or automatic thru the shop interface (or any other), there is no real reason why this should be different. Second Principle, everything should be logged.

Stock Changes: 
Order status changed from Pending payment  / Cancelled to Processing / Completed / On-Hold should decrease stock.
Order status changed from Processing / Completed / On-Hold to Pending payment  / Cancelled should increase stock.
Editing order on the back end, add a product, stock is decreased, remove a product, stock is increased.

If there are users who want to make order status changes or order item changes and don't want the stock levels to change then there can be a Yes / No message box asking if the user want to see those changes reflected in the stock. I cannot imagine one situation where this can be needed, but we never know.

**Comment by kloon** [comment]
> Any rational why this was removed?

It was removed because the reduction and increase now happen automatically based on the normal order status flow as per front-end orders. This was requested by many users who did not want to manually reduce stock for manual orders.

**Comment by filpet** [comment]
Totally agree with @negapo ! 

Firstly, we face the problem that orders in "On-Hold" do not reduce the stock, however all our orders go to "On-Hold" automatically as we have to control all of them for our specific customer group before putting them into "Processing". Our customers pay by a monthly bill after getting the products, there is no payment gateway. So all of a sudden all our orders are made without reducing stock levels!!!

Secondly we have to edit quite a lot of our weekly orders after them being in "Processing" and so on.

Please go back to allowing more flexible edits within orders and also give users the option to choose which order status influences stock! We really need the "On Hold" orders to do so!

Thanks a lot!

**Comment by rekkitatu** [comment]
We are occasionally seeing situations where Woocommerce incorrectly reports the quantities changed after payment in the order notes sidebar. Although all our products are unique (~30000 separate products), so they never have stock quantities more than 1. All these products even have the sold individually feature enabled.  Still sometimes Woocommerce reports in the order notes that the quantities are changed from 2 to 1 which shouldn't be possible. Then the products view show that the product has quantity of 0 with the stock status "instock" which shouldn't be happening either. This typically happens with large orders that have more than 10 separate items. Using the increase/reduce stock feature from the edit order page has been the quickest way to fix these cases.

There aren't any errors in the log files related to these issues. We haven't been able to find any fix to this problem. We hoped that maybe the this new WC 3.5.x would somehow fix the issue with the improved stock handling but just recently I saw an order where the stock reduction happened incorrectly. Just wish that the upcoming custom product tables feature would finally improve the performance and perhaps fix this also. Until then I think the ability to make manual changes to the stock from the order view is paramount.

**Comment by Matiasi** [comment]
I don't understand why this was removed. You can make it update automatically but still preserve the option of manually editing. This specially important as Woo doesn't have a decent POS option for handling manual orders

**Comment by moneytones** [comment]
Yes, this feature should have never been removed.

Examples: Customer wants to add/remove items form order. Customer wants to change his order. 

Now we have to got the product page and edit the inventory and it's lost in the notes so we can't track if the item was reduced manually or not. Makes order fulfillment a lot less efficient in a real word situation.

PLEASE PLEASE BRING IT BACK.

**Comment by urkaGR** [comment]
PLEASE bring the feature back!!! 
Examples: 
Customer wants to add/remove items form order. Customer wants to change his order.

But I also have a problem with the new feature who bring all order items back to stock if you change order in status "waiting for payment"
Regular Customers need quick support and we send items exceptionally before payment is on our bank.
So we have send it but not finished because waiting for payment and now the system fill up the stock with my sent products ?!?!

items automatic book back on stock should only be on status "Storno" i think!
On Refund also maybe a guarantee so item come not back or a lost parcel or something else so Admin have to decide if product goes back to stock or not!

**Comment by rrennick** [comment]
@thpwindsor The PR above is in the 3.5.3 milestone.

**Comment by oblaum** [comment]
also the automatic version at the moment does not handle the scenario:
when i delete a product from the order, the stock is not changing as well.

**Comment by Anarchrist** [comment]
Please bring this feature back

**Comment by iSevenDays** [comment]
Please bring this feature back!

It is really annoying when editing orders. I have to go to each product, increase/decrease stock and write comments to the order that I did something in order to not forget.

Another use case:
A user placed an order with coin X => stock is decreased for coin X by 1.

The user calls me and asks to edit the order and add 1 more coin X, and coin Y.

I have to go to product coin X, decrease stock by 1, write a comment about my action.
Then I need to go to coin Y, decrease stock by 1 and also write a comment.

Because if I have another call where the user asks me to change something, or cancel the order - I will forget everything I did without comments.

**Comment by thpwindsor** [comment]
> @thpwindsor The PR above is in the 3.5.3 milestone.

Hello @rrennick . This issue doesn't appear to have been solved in the latest update 3.5.3. Has it been delayed to 3.5.4 or do we have to manually change the code based on your pull request if we want this feature back?

**Comment by melindahelt** [comment]
@thpwindsor and @Xenokorp The 3.5.3 update was a quick one to fix a WordPress compatibility issue as outlined at https://woocommerce.wordpress.com/2018/12/20/woocommerce-3-5-3-release-notes/, the fix for this issue should still be included in a future release.

**Comment by thpwindsor** [comment]
@melindahelt Thanks for the prompt reply and update. Could you possibly let us know on which future update we can expect this feature to return?

**Comment by rrennick** [comment]
> Could you possibly let us know on which future update we can expect this feature to return?

It was moved to 3.5.4. The link to the PR is in the comment thread above.

**Comment by mikejolley** [comment]
There are several related, but not the same, requests in this thread and not all will be solved by adding back those buttons. Probably because not everyone understands the request :)

> No idea what it means, but it seems positive so we get that back :)

These problems essentially include:

1. Order edits - deleting and adding items to orders do not adjust stock regardless of order status meaning it must be done manually the slow way.

> I can give you a very simple example. A costumer places an order for a red tshirt and pays it, it then emails me asking to change it to a blue tshirt and then all hell breaks loose. I have to remove the red tshirt, go to the product page, increase the stock by one unit, then add the blue tshirt to the order, go to the product page and decrease the inventory by one unit. And nothing of this gets logged.

>  I always need to edit an order due to customer requests. 

2. Stock changes that should happen, are not happening. 

> we face the problem that orders in "On-Hold" do not reduce the stock, however all our orders go to "On-Hold" automatically as we have to control all of them for our specific customer group before putting them into "Processing".

> all orders I shipped yesterday got all the stock reduced twice (once by me when placed, once by the system upon completion).

These 2 scenarios sound like conflicts tbh. Those events are already handled on status change. Whilst the controls would make fixing these easier, it doesn't solve the cause so doesn't justufy adding the buttons back IMO.

3. Mis-use of order statuses (@urkaGR):

> But I also have a problem with the new feature who bring all order items back to stock if you change order in status "waiting for payment"
> Regular Customers need quick support and we send items exceptionally before payment is on our bank.

This is exactly the scenario on-hold status is designed for. The order is on hold until payment comes in. This is the status BACS gateway uses.

---

@ben72 for your list:

> When picking the order it's a great way to keep track of progress
> Put a returned item back into stock when for instance changing to another size and get the action logged at the same time.

See point 1 above.

> Remove an item from stock if replacing a faulty item and get the action logged at the same time.

If stock was managed when adding an item this could be achived; you'd add the replacement to the order leaving the faulty item line in place so there is a record.

> Put items back in stock when pre-paid order is canceled.

Cancelled orders do restore stock as-is.

--- 

Thanks @rrennick for the PR restoring this but I don't think thats a good way forward. The mix of manual and automatic stock reduction events adds confusion i.e. who should be doing what, when, what has been done already (think large orders), not to mention various user interviews showing those buttons are not even well known due to the hidden UI elements.

 If we can avoid them, and keep things automated, that would be a win.

[This is closer to what we're aiming for.](https://github.com/woocommerce/woocommerce/issues/21754#issuecomment-436644841)

Right now we do increases/decreases based on order status. This works for orders via checkout fine.

The main problem seems to stem from manual orders; admin setting up an order for a user to pay without stock being mishandled, and order edits which see line items changing at some point during the order flow.

I think the best way to tackle this would be to:

1. Restore stock (if it's been reduced) when deleting a line item from an order.
2. Remove stock when adding line items to an order (any order) manually.
3. Adjust stock when editing line item qty on an order manually.

All 3 events above should add a note to track changes.

For the OP, @melindahelt that looks like a manual ordering scenario which would be helped by the above.

I'll take a stab at a PR to see how/if this will function as planned.

cc @claudiulodro

**Comment by mikejolley** [comment]
See https://github.com/woocommerce/woocommerce/pull/22329 - there are some gifs in there showing usage.

**Comment by filpet** [comment]
> See #22329 - there are some gifs in there showing usage.

Would love that! Can't wait for the release!

**Comment by rrennick** [comment]
Note that my PR has a commit for #21837.

**Comment by windsorraw** [comment]
In response to #22329, not sure if I should post this here or there so I did both:

The 3 examples above appear to cover off all the scenarios where we would need to manually increase or decrease individual line items and this suggested solution would achieve what is needed and in fact make things easier with less steps for the user, so thank you for going thr extra mile instead of reverting back. Always moving forward!

Here is one scenario that come up regularly on my site that this currently would not solve:

Order is placed online for 15 items for pick up in store. Some items are in stock and some are not. In such a case we change the order status to "on hold", which to the customer means that their order will not be available untill the items come in. We would normally increase all line items so the items go back in stock can be sold to others.

When it comes time to place our weekly restock order, we would increase line items on the order and change to "processing" which would update my inventory to show what needs to be ordered. If changing order statuses would increase/decrease the entire order this would be great but currently this is not the case.

If an order has 25 items, manually reducing and increasing inventory is a major pain with a lot of room for human error.

**Comment by mikejolley** [comment]
@windsorraw I'm not sure how that would be best covered. It sounds like i that were to happen, you could risk selling other items in the order instead of leaving it allocated to the user? On-hold is correct status, but changing those stock levels feels off.

**Comment by windsorraw** [comment]
@mikejolley You nailed it, that's just it. We want to be able to sell those items to others as this customer will be waiting for our next delivery anyways and we can order any items our of stock for them. Please follow this simplified example.

Order:
2 x red shirt (current stock is 4 red shirts)
2 x blue shirt (current stock is 1 blue shirt)

Once reduced stock would:
2 red shirts
-1 blue shirts. 
However, the customer wants to pick up their whole order at once and will wait for it all to come in (which is almost always the case for our business).

Now we still have:
4 red shirts
2 blue shirts 
available to sell this week so I have to increase the inventory. Assuming we did that and we sold out. At the end of the week we change all "on hold" orders to "processing" and reduce inventory. 
Now we have 
-2 red shirts
-2 blue shirts
And I would order accordingly.

Does this makes sense? We have been doing it this way for over 4 years, let me know if im not explaining it well.

**Comment by urkaGR** [comment]
3.	Mis-use of order statuses (@urkaGR <https://github.com/urkaGR> ):

Hi, the problem of this status is “on hold” is, that the items are back as available in the shop not reserved for this order

we are only waiting for the payment when we send with invoice– items are on the way to the customer and not available for other customers in shop.

the item stock does not depend of the payment, it dempends of an order or storno something.

we had made now a quick and dirty issue and added a new status “delivery” – so we know, that these are delivered and we are waiting for the payment. if we get the payment, then we set on finish.

Von: Mike Jolley [mailto:notifications@github.com] 
Gesendet: Donnerstag, 3. Jänner 2019 11:47
An: woocommerce/woocommerce <woocommerce@noreply.github.com>
Cc: urkaGR <urka@gmx.net>; Mention <mention@noreply.github.com>
Betreff: Re: [woocommerce/woocommerce] Need option to manually reduce stock on order level (#21754)

 

There are several related, but not the same, requests in this thread and not all will be solved by adding back those buttons. Probably because not everyone understands the request :)

No idea what it means, but it seems positive so we get that back :)

These problems essentially include:

1.	Order edits - deleting and adding items to orders do not adjust stock regardless of order status meaning it must be done manually the slow way.

I can give you a very simple example. A costumer places an order for a red tshirt and pays it, it then emails me asking to change it to a blue tshirt and then all hell breaks loose. I have to remove the red tshirt, go to the product page, increase the stock by one unit, then add the blue tshirt to the order, go to the product page and decrease the inventory by one unit. And nothing of this gets logged.

I always need to edit an order due to customer requests.

2.	Stock changes that should happen, are not happening.

we face the problem that orders in "On-Hold" do not reduce the stock, however all our orders go to "On-Hold" automatically as we have to control all of them for our specific customer group before putting them into "Processing".

all orders I shipped yesterday got all the stock reduced twice (once by me when placed, once by the system upon completion).

These 2 scenarios sound like conflicts tbh. Those events are already handled on status change. Whilst the controls would make fixing these easier, it doesn't solve the cause so doesn't justufy adding the buttons back IMO.

3.	Mis-use of order statuses (@urkaGR <https://github.com/urkaGR> ):

But I also have a problem with the new feature who bring all order items back to stock if you change order in status "waiting for payment"
Regular Customers need quick support and we send items exceptionally before payment is on our bank.

This is exactly the scenario on-hold status is designed for. The order is on hold until payment comes in. This is the status BACS gateway uses.

  _____  

@ben72 <https://github.com/ben72>  for your list:

When picking the order it's a great way to keep track of progress
Put a returned item back into stock when for instance changing to another size and get the action logged at the same time.

See point 1 above.

Remove an item from stock if replacing a faulty item and get the action logged at the same time.

If stock was managed when adding an item this could be achived; you'd add the replacement to the order leaving the faulty item line in place so there is a record.

Put items back in stock when pre-paid order is canceled.

Cancelled orders do restore stock as-is.

  _____  

Thanks @rrennick <https://github.com/rrennick>  for the PR restoring this but I don't think thats a good way forward. The mix of manual and automatic stock reduction events adds confusion i.e. who should be doing what, when, what has been done already (think large orders), not to mention various user interviews showing those buttons are not even well known due to the hidden UI elements.

If we can avoid them, and keep things automated, that would be a win.

This is closer to what we're aiming for. <https://github.com/woocommerce/woocommerce/issues/21754#issuecomment-436644841> 

Right now we do increases/decreases based on order status. This works for orders via checkout fine.

The main problem seems to stem from manual orders; admin setting up an order for a user to pay without stock being mishandled, and order edits which see line items changing at some point during the order flow.

I think the best way to tackle this would be to:

1.	Restore stock (if it's been reduced) when deleting a line item from an order.
2.	Remove stock when adding line items to an order (any order) manually.
3.	Adjust stock when editing line item qty on an order manually.

All 3 events above should add a note to track changes.

For the OP, @melindahelt <https://github.com/melindahelt>  that looks like a manual ordering scenario which would be helped by the above.

I'll take a stab at a PR to see how/if this will function as planned.

cc @claudiulodro <https://github.com/claudiulodro>

**Comment by mikejolley** [comment]
@windsorraw It sort of makes sense yes but it doens't sound like a common enough case to introduce back special tools to handle?

@urkaGR I don't know why but on-hold orders do reserve stock. As soon as order goes on-hold, stock reduction takes place.

**Comment by windsorraw** [comment]
@mikejolley when the order goes on hold right now the stock goes "on hold" but the inventory is not increased. Would it be a possibility to give us an option to control what happens to inventory in particular ordr statuses?

**Comment by urkaGR** [comment]
Hi,
the status is not "on-hold" it is "pending payment"
if we send out an order but don't have the payment because customer get an invoice we set the status pending payment before update. but this status put back the items in product stock in new version.
sorry, there was an missunderstand lost in translation.

the status "on hold" is correct. but this means, that I have done nothing with the order - it is only "on hold".

 

 


Gesendet: Donnerstag, 03. Januar 2019 um 18:21 Uhr
Von: "Mike Jolley" <notifications@github.com>
An: woocommerce/woocommerce <woocommerce@noreply.github.com>
Cc: urkaGR <urka@gmx.net>, Mention <mention@noreply.github.com>
Betreff: Re: [woocommerce/woocommerce] Need option to manually reduce stock on order level (#21754)


@windsorraw It sort of makes sense yes but it doens't sound like a common enough case to introduce back special tools to handle?

@urkaGR I don't know why but on-hold orders do reserve stock. As soon as order goes on-hold, stock reduction takes place.

**Comment by omidcg** [comment]
Hi. Please back reduce/increase stock in order level.

**Comment by davidlonguk** [comment]
"normal order status flow" - so the assumption is that the status flow is the same for every store owner?? What about custom order statuses and order flows that do not match what Woo consider "normal". 
It sounds very much like a Steve Jobs "your holding it wrong" answer.

**Comment by AKSUTAH** [comment]
We need this feature for making changes to orders. If a customer calls and wants to remove or replace an item, this feature is used for returning the removed or replaced items back to inventory.

**Comment by zsebacsi** [comment]
+1 for re-adding this useful feature.
it eases work when modifying orders manually.

**Comment by PIonut09** [comment]
+1 to re-add this feature

**Comment by davsev** [comment]
+1 Any news? this thread is open for 6 months and still no solution? is there any walkaround?

**Comment by tommydamani** [comment]
This is so annoying....  just bring the damn feature back what the hell??? Can't you see we use this function for whatever reason? You people constantly change and remove things that pisses us off....

We create our workflow based on what we have to work with and then when you remove something THAT WE USE EVERY DAY, it stuffs us up. What freaking harm is it to leave the damn feature there?????? You can't see we use it?????

**Comment by xpd** [comment]
+1

**Comment by Web-developer-Alexander** [comment]
Hello all!
+1 this option is very necessary! We currently have WooCommerce version: 3.4.7 and we do not want to update the plugin to the latest version in order not to lose manual inventory management.

**Comment by AKSUTAH** [comment]
Another example of when this needs to be used is if a customer wants to exchange things on an order. Maybe they ordered ten 1 packs and I want to change it to 1 ten pack. I'd need to manually change the stock on the exchanged items.

**Comment by moneytones** [comment]
This is still not FIXED????
ᐧ

**Comment by tommydamani** [comment]
We are STILL NOT UPDATING the plugin because we use that feature all the fkn time. Why the hell would you remove it? What difference does it make???????????????

**Comment by windsorraw** [comment]
Woocommerce team has worked hard at sorting out the increasing and decreasing inventory issues. Many of the issues since removing the manual feature have been fixed but there are still some gaps.

A BIG problem we are having is when we change an order status to "on hold" we need the inventory to be returned (increased) for the entire order. Currently the order statuses are set-up to decrease when the order is placed and increase when the order is cancelled. 

Would it be possible to put in a feature where not just a cancelled order but an "on hold" order would also increase the inventory? Is there a simple line of code that I can manually go in and edit for this?

Thank You

**Comment by EdithAllison** [comment]
> Would it be possible to put in a feature where not just a cancelled order but an "on hold" order would also increase the inventory? Is there a simple line of code that I can manually go in and edit for this?

There are actions. Have a look in /woocommerce/includes/wc-stock-functions.php . By adding / removing actions from a status you can control when stock gets reduced / increased. You'd put the actions into the functions.php file of your child theme. 

add_action( 'woocommerce_order_status_xxx', 'wc_maybe_reduce_stock_levels' );
add_action( 'woocommerce_order_status_xxx', 'wc_maybe_increase_stock_levels' );

where xxx is the status name

**Comment by mattybri** [comment]
I've found a solution! In your FTP, go to:  /wp-content/plugins/woocommerce/includes/wc-stock-functions.php and remove code from line 78 - 109. 
This is the code which tells WooCommerce to lower stock level after order.
Now when you order a product, the stock level does't change and can still be edited manually in Product settings

**Comment by nikaslg** [comment]
Please return the feature...

**Comment by davidcardonacm** [comment]
Hi, why dont you simple add this as an optional check mark in products setting?
In this way, they one's who need it would have it, the ones who dont will leave it as default,
Thanks.

**Comment by babylines** [comment]
> Closing in favour of #22329 which is targetted for 3.6.0 (may make some tweaks but it's otherwise working).

**The problem is not a bug. The problem is some store owners do not want stock automatic adjustments to be forced to match the order changes. This thread is riddled with examples. Please do the professional and correct thing and give an option.**

/home/greenz/public_html/wp-content/plugins/woocommerce/includes/admin/wc-admin-functions.php

			$changed_stock = wc_maybe_adjust_line_item_product_stock( $item );

Just add a setting to wc or a selectbox to the edit order page and then add the condition to the above line. (Or a user can just comment it out like I did =)

**Comment by ecottles** [comment]
The politics of order statuses. What the community wants it gets.  I agree with babylines. All statuses should be in control of developer to define according to the business model they are serving. However, and unfortunately the politics of order statuses wins the day!
